src/groupNoble.ts

```diff
@@ -0,0 +1,338 @@
+import { Field as FieldCons } from '@noble/curves/abstract/modular'
+import { bytesToNumberBE } from '@noble/curves/abstract/utils'
+import { expand_message_xmd } from '@noble/curves/abstract/hash-to-curve'
+import { ProjConstructor, ProjPointType } from '@noble/curves/abstract/weierstrass'
+import { CHash } from '@noble/hashes/utils'
+import {
+    Deserializer,
+    Elt,
+    getGroupID,
+    Group,
+    GroupCons,
+    GroupID,
+    GroupIDs,
+    Scalar
+} from './groupTypes'
+
+export type Field = ReturnType<typeof FieldCons>
+
+function errDeserialization(T: { name: string }) {
+    return new Error(`group: deserialization of ${T.name} failed.`)
+}
+
+function errGroup(X: GroupID, Y: GroupID) {
+    return new Error(`group: mismatch between groups ${X} and ${Y}.`)
+}
+
+function errBadGroup(X: string) {
+    return new Error(`group: bad group name ${X}.`)
+}
+
+function compat(x: { g: Group }, y: { g: Group }): void | never {
+    if (x.g.id !== y.g.id) throw errGroup(x.g.id, y.g.id)
+}
+
+export class ScalarNb implements Scalar {
+    private readonly fp: Field
+    public readonly k: bigint
+
+    private constructor(public readonly g: GroupNb, k: bigint) {
+        this.fp = getCurve(this.g.id).fp
+        this.k = this.fp.create(k)
+    }
+
+    static new(g: GroupNb): ScalarNb {
+        return new ScalarNb(g, BigInt(0))
+    }
+
+    isEqual(s: ScalarNb): boolean {
+        return this.k === s.k
+    }
+
+    isZero(): boolean {
+        return this.k === BigInt(0)
+    }
+
+    add(s: ScalarNb): ScalarNb {
+        compat(this, s)
+        return new ScalarNb(this.g, this.fp.add(this.k, s.k))
+    }
+
+    sub(s: ScalarNb): ScalarNb {
+        compat(this, s)
+        return new ScalarNb(this.g, this.fp.sub(this.k, s.k))
+    }
+
+    mul(s: ScalarNb): ScalarNb {
+        compat(this, s)
+        return new ScalarNb(this.g, this.fp.mul(this.k, s.k))
+    }
+
+    inv(): ScalarNb {
+        return new ScalarNb(this.g, this.fp.inv(this.k))
+    }
+
+    serialize(): Uint8Array {
+        const k = this.fp.create(this.k)
+        const ab = this.fp.toBytes(k)
+        const unPadded = new Uint8Array(ab)
+        const serScalar = new Uint8Array(this.g.size)
+        serScalar.set(unPadded, this.g.size - unPadded.length)
+        return serScalar
+    }
+
+    static size(g: GroupNb): number {
+        return g.size
+    }
+
+    static deserialize(g: GroupNb, bytes: Uint8Array): ScalarNb {
+        checkSize(bytes, ScalarNb, g)
+        const array = bytes.subarray(0, g.size)
+        const k = bytesToNumberBE(array)
+        const fp = getCurve(g.id).fp
+        if (k >= fp.ORDER) {
+            throw errDeserialization(ScalarNb)
+        }
+        return new ScalarNb(g, k)
+    }
+
+    static hash(g: GroupNb, msg: Uint8Array, dst: Uint8Array): ScalarNb {
+        const { hash, L } = getHashParams(g.id)
+        const s = expand_message_xmd(msg, dst, L, hash)
+        return new ScalarNb(g, bytesToNumberBE(s))
```

Comment:
here there is no reduction modulo group order, or better invoke `hash_to_field`

Reply:
added a modulo by group order

Reply:
> better invoke hash_to_field

btw, I had a quick look into that, but wasn't able to get the tests to pass. 
Will look again later

Reply:
```typescript
    static async hash(g: GroupSj, msg: Uint8Array, dst: Uint8Array): Promise<ScalarSj> {
        const { hash, L } = getHashParams(g.id)
        const bytes = await expandXMD(hash, msg, dst, L)
        const array = Array.from(bytes)
        const bitArr = sjcl.codec.bytes.toBits(array)
        const k = sjcl.bn.fromBits(bitArr).mod(getCurve(g.id).r)
        return new ScalarSj(g, k)
    }
```

Actually, the sjcl implementation simply uses expandMD ?

Reply:
@paulmillr @armfazh

Ok, I updated it to use hash_to_field

I noticed something that I need to understand better. 
hashToCurve -> element uses hash_to_field with p set (by @noble, see [here](https://github.com/paulmillr/noble-curves/blob/c2edc978683c1a688eecc0f020a80a1ff8ded749/src/p256.ts#L11-L41)) to the curve's prime, whereas we seem to need to use the sub group order as p for the tests to pass here.

Reply:
Wait, p is the size of the field of possible values for x,y in an element ?
Whereas n is size of of the field of possible elements, thus suitable for scalar multiplication  ?

Reply:
In an elliptic curve there are two important primes:
p, the prime used to define the curve params and the (x,y,z) coordinates.
n, (or r), the prime that defines the group order. Equiv, defines the range for scalars.

Since both are primes, you can always define a prime field (GF) with each of them. 

For hash_to_curve, we use internally `hash_to_field( prime = p)` (i.e., setting the prime to p) for getting candidates of coordinates.

However, for random_scalar, we can use internally `hash_to_field( prime = n )` (i.e., setting the prime to n), so we can get scalars in such a range.

Hope this clarifies.

Reply:
Thanks! Yeah, that makes sense